homeassistant/components/tfl/common.py

```diff
@@ -0,0 +1,34 @@
+"""Common utility functions for the Transport for London integration."""
+
+from collections.abc import Callable
+from urllib.error import HTTPError, URLError
+
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import HomeAssistantError
+
+
+async def call_tfl_api[*_Ts, _T](
+    hass: HomeAssistant, target: Callable[[*_Ts], _T], *args: *_Ts
+) -> _T:
+    """Execute a call to TfL using the tflwrapper library using common error handling."""
+    try:
+        return_val = await hass.async_add_executor_job(target, *args)
+    except HTTPError as exception:
+        # TfL's API returns a 429 if you pass an invalid app_key, but we also check
+        # for other reasonable error codes in case their behaviour changes
+        error_code = exception.code
+        if error_code in (429, 401, 403):
+            raise InvalidAuth from exception
+        raise
+    except URLError as exception:
+        raise CannotConnect from exception
+
+    return return_val
+
+
+class CannotConnect(HomeAssistantError):
+    """Error to indicate we cannot connect."""
+
+
+class InvalidAuth(HomeAssistantError):
+    """Error to indicate there is invalid auth."""
```

Comment:
The library is providing just a thin wrapper around the TfL REST API, so doesn't make an attempt to interpret any error codes - it just lets them bubble up. From the library perspective, that's a design choice.

From the integration perspective, it's reasonable to hedge our bets slightly as the choice of which 4xx error code is returned by the TfL REST API for invalid auth isn't great (429) - but perhaps in future that changes to 401 and if so we don't want to have to do a bug fix and update the integration.
And that then drives centralising the logic here rather than duplicating it at all the call sites.

The library wants to keep it simple and not assume that 429 is invalid auth and have to release a bug fix if the TfL REST API starts using 429 properly.
I think the library is making reasonable choices.

But yeah - ideally this code wouldn't need to exist.